test/sandbox/test/rendering_test.rb

```diff
@@ -1078,4 +1078,30 @@ def test_renders_nested_collection
       index += 1
     end
   end
+
+  def test_concurrency_deadlock
+    with_compiler_mode(ViewComponent::Compiler::DEVELOPMENT_MODE) do
+      with_new_cache do
+        mutex = Mutex.new
+
+        t1 = Thread.new do
+          mutex.synchronize do
+            sleep 0.02
+            render_inline(ContentEvalComponent.new)
+          end
+        end
+
+        t = Thread.new do
+          render_inline(ContentEvalComponent.new) do
+            mutex.synchronize do
+              sleep 0.01
+            end
+          end
+        end
+
+        t1.join
+        t.join
```

Comment:
There's no assertion here, since the old code causes an error that looks like this:

```
Error:
RenderingTest#test_concurrency_deadlock:
fatal: No live threads left. Deadlock?
7 threads, 7 sleeps current:0x000000010697e260 main thread:0x00000001516046f0
* #<Thread:0x000000015205fae8 sleep_forever>
   rb_thread_t:0x00000001516046f0 native:0x00000001031ac580 int:0
   /Users/blake/Code/view_component/test/sandbox/test/rendering_test.rb:1103:in `join'
```

followed by a lot of stacktrace information 😄